src/index.ts

```diff
@@ -30,24 +32,39 @@ export class VisualObserver {
 
   #elements = new Map<Element, VisualObserverElement>();
 
-  #resizeObserver = new ResizeObserver((entries: ResizeObserverEntry[]) => {
-    const visualEntries: VisualObserverEntry[] = [];
-
+  #resizeObserver = new ResizeObserver((entries) => {
+    const rootEntry = entries.find((entry) => entry.target === this.#root);
     // Any time the root element resizes we need to refresh all the observed elements.
-    if (entries.some((entry) => entry.target === root)) {
+    if (rootEntry !== undefined) {
+      this.#rootRect = rootEntry.contentRect;
+
       this.#elements.forEach((_, target) => {
         // Why force a refresh? we really just need to reset the IntersectionObserver?
-        visualEntries.push(this.#refreshElement(target));
+        this.#appendEntry(this.#refreshElement(target));
       });
     } else {
       for (const entry of entries) {
-        visualEntries.push(this.#refreshElement(entry.target, entry.contentRect));
+        this.#appendEntry(this.#refreshElement(entry.target));
       }
     }
-
-    this.#callback(visualEntries, this);
   });
 
+  #appendEntry(entry: VisualObserverEntry) {
+    // deduplicate the same target
+    this.#entries.push(entry);
+
+    if (!this.#rafId === null) {
```

Comment:
is this a typo? `!`...

Also, general advice, make this zero by default as V8/friends will optimize for object shape. `null` => `number` (effectively an int) is "slow".
(This is the tiniest micro optimization ever, but just some advice.)

Reply:
oof it is a typo!